src/persist/src/operators/stream.rs

```diff
@@ -463,6 +466,200 @@ where
+        // don't have to wait until the allowed frontier advances past a given frontier, we only
+        // need to be careful not to advance the compaction frontier up to the seal frontier, for
+        // which the input frontier is a proxy.
+        let mut candidate_compaction_frontier: Antichain<u64> =
+            Antichain::from_elem(timely::progress::Timestamp::minimum());
+
+        op.build_reschedule(move |_capabilities| {
+            move |frontiers| {
+                let mut data_output = data_output.activate();
+
+                // Pass through all data.
+                data_input.for_each(|cap, data| {
+                    data.swap(&mut data_buffer);
+
+                    let mut session = data_output.session(&cap);
+                    session.give_vec(&mut data_buffer);
+                });
+
+                if !active_operator {
+                    // We are always complete if we're not the active operator.
+                    return false;
+                }
+
+                let allowed_compaction_frontier = allowed_compaction_frontier.borrow_mut();
+                let input_frontier = frontiers[0].frontier();
+
+                effective_input_frontier.clear();
+                effective_input_frontier.extend(input_frontier.iter().cloned());
+                effective_input_frontier.extend(allowed_compaction_frontier.iter().cloned());
+
+                if PartialOrder::less_than(
+                    &candidate_compaction_frontier,
+                    &effective_input_frontier,
+                ) {
+                    // We can now safely compact to this candidate frontier, because we know the
+                    // input frontier combined with the allowed compaction is strictly beyond (that
+                    // is >).
+                    std::mem::swap(&mut compaction_frontier, &mut candidate_compaction_frontier);
+
+                    // The current combined input frontier becomes the next candidate up to which
+                    // we could allow compaction, once it is strictly dominated by the then current
+                    // effective input frontier.
+                    candidate_compaction_frontier.clear();
+                    candidate_compaction_frontier.extend(effective_input_frontier.iter().cloned());
+
+                    tracing::trace!(
+                        "In {}, allowing compaction up to {:?}. Effective input frontier: {:?}",
+                        &operator_name,
+                        compaction_frontier,
+                        effective_input_frontier
+                    );
+
+                    let future = write.allow_compaction(compaction_frontier.clone());
+
+                    pending_futures.push_back(CompactionFuture {
+                        frontier: compaction_frontier.clone(),
+                        future,
+                    });
+                }
+
+                // Swing through all pending futures and see if they're ready. Ready futures will
```

Comment:
Not for this PR, but really seems like a bunch of these operator impls have a pretty common pattern that we could maybe extract.